### ISSUE TITLE
Blocking when running commands looking for stdin close

### ISSUE THREAD

**Issue by plockc** [pyinvoke/invoke#552]
Example:
```
@task
def cat(c):
    c.run("cat")
```
```
echo Hi | timeout 3 invoke cat || echo failed
```
This prints Hi but never exits because cat never receives a close of stdin and is waiting for more, and invoke is waiting on the subprocess.

In a different issue, I notice that the console has trouble after running the above command 

Caused by two bugs:

First is that ExceptionHandlingThread.is_dead return expression is wrong, it should be "or" instead of "and":
https://github.com/pyinvoke/invoke/blob/master/invoke/util.py#L274

Second, is that the stdin for the process is not closed after receiving EOF:
https://github.com/pyinvoke/invoke/blob/master/invoke/runners.py#L659

Python 3.6.5
Ubuntu 18.04
Invoke 1.0.0.0

**Comment by bitprophet** [comment]
Can confirm the lack of exiting easily enough.

Not getting terminal weirdness however, even under bash, which is where I am able to reproduce the potentially-related fabric/fabric#1812. Still trying to determine if the two issues stem from the same bugs.

**Comment by bitprophet** [comment]
Still ongoing but that Fabric ticket seems to be caused by executing `with character_buffered(...)` multiple times concurrently (such that all but the first invocation erroneously snapshots, then restores, an already-character-buffered terminal state) which is what causes the weird local terminal behavior (specifically, the `echo` flag is set to off, and then never properly restored).

Given the code is a contextmanager, I don't see why it would ever hit that particular issue in single-threaded Invoke-only code; no matter what happens inside the block (early exit, error, process termination short of SIGKILL) the post-yield, end-of-block terminal restoration code (which is in a `finally` clause) should always run.

In addition, when (our) stdin is hooked up to a pipe, as it is here, that contextmanager should become a no-op. The detection of whether to do that has historically been a bit fragile, so maybe this exposes another hole in it. Either way, I can't reproduce that angle, but I can reproduce the hanging/never-exiting problem so I'll focus on that first and we can make a new issue for terminal state corruption if necessary.

**Comment by plockc** [comment]
hi @bitprophet, the terminal weirdness only happened once, I can't reproduce, I probably should not have mentioned in this ticket.

I have locally ran 1.1.1 and still can reproduce the blocking issue, while taking my 1.0.0 patch and rebasing on top of 1.1.1 works for the "cat" test.

I added test cases and a discussion about why I think the test cases were wrong too, mind looking at my comments and the test case updated assertions?
https://github.com/pyinvoke/invoke/pull/553

**Comment by bitprophet** [comment]
Think I'm hitting this internally now and traced it to something similar-but-different - a lack of submitting EOF down the pipe when non-mirrored in_stream hits its end.

I suspect that the explicit file close in OP's commit (https://github.com/plockc/invoke/commit/4d6ab1143ae998bc1bc97862ac40974716511d5e) will do the same thing and will test that momentarily.

FWIW my own fix of `Runner.send_interrupt()` results in a later thread deadlock / logic hang (we hit the `or` logic in `Runner.wait`, where `dead_threads` is True but `proc_finished` is still False) so if that still pops up after switching to the file close approach, more work will be needed.

**Comment by bitprophet** [comment]
When looking at extending that commit to Fabric (which is where I am troubleshooting this for now) I notice we have a mostly-noop `Runner.stop` method. Thinking maybe we want to fold `_close_proc_stdin` into that method (or at least call the latter from the former, then call the former where the patch currently calls the latter...). Still figuring out if this even fixes the whole problem for me though.

**Comment by bitprophet** [comment]
Does appear to fix my issue, however there's some race condition where maybe 10% of the time, the  remote end (in this case, a simple `wc -l`) exits `-1` instead of 0. Gotta remember exactly what that means, though my offhand guess is it means we're closing down before the remote end is and it's getting a broken pipe.

**Comment by bitprophet** [comment]
Interestingly, the bad exit code happens even when the remote program doesn't touch its stdin. I also realize...there's no stdout coming back down the pipe. So something's very clearly wrong here. Removing `in_stream` from the equation makes everything normal again, even with this patch involved.

The issue with lack of stdout may be a separate, related logic error triggered by `in_stream` though.

**Comment by bitprophet** [comment]
Thinking this may just be some sort of hilarity where we read a small file from stdin and it gets read/written so fast it's happening before the remote end is even set up correctly; that in tandem with the changed termination logic may mean we close things down before the remote end has even started execution.

That would explain this behavior perfectly. Plus given we clearly haven't tested the non-interactive stdin scenario anywhere near as much as the interactive-stdin scenario, it makes sense we never tripped over this before - in the other scenario, stdin will never exit before the remote end, because it's a live terminal.

**Comment by bitprophet** [comment]
The negative-1 exit status I am getting smells like it may be SSH specific so this is going to be another spot where Fabric bleeds down into Invoke, grump. Does explain why the OP figured the patch worked OK for them. Seems plausible that we'll end up merging this partly as-is and having to address something else on top of it in Fabric's runner subclass.

**Comment by bitprophet** [comment]
Yea, as always in this arena, there are too many damned things all competing. When I get the scenario working for a non-reading-from-stdin subprocess, the reading-from-stdin scenario breaks again, and vice versa.

I thought I had something workable where Fabric was just treating `Runner._close_proc_stdin` as a noop - because Paramiko/SSH have no real concept of closing just stdin, only the entire overall channel - was working ok in a base case (whereas I was closing the whole channel before, like an idiot, and that causes its own problem chain).

But as soon as we return to the real scenario under test - a program actually trying to follow stdin - we're back to square one where that program seems to hang out forever waiting for stdin to close.

**Comment by bitprophet** [comment]
Wow, Paramiko's had a stdin-closing related ticket since at least 2014. Not sure why it's never surfaced in my knowledge. Love software. It's paramiko/paramiko#322. Which hints at a Channel member I didn't even remember existed, `shutdown_write`.

When I dig, it does indeed do what I expected from grepping around for "EOF" in Paramiko - it sends an explicit SSH EOF message type. And plugging THAT into Fabric's version of our new method, seems to make things Just Work.

I'd say hallelujah but I'm sure I'm overlooking something and this isn't over yet, because it is that kind of day.

**Comment by bitprophet** [comment]
Also, we still need the usual administrivia here - tests, changelog etc. The tests will probably be an enormous pain in the ass and my tolerance at this point is at an all time low, so I'll probably just do something half-baked.

**Comment by bitprophet** [comment]
Revisiting this so it can get closed out. Writing some quick tests & they're exposing a flaw with my reasoning in the commit that removes the `break` after sending EOF down the pipe: yes, we do need to avoid that race condition, but just letting the loop keep rolling causes the EOF to be re-sent hundreds or thousands of times! (as evidenced by my test naively assuming that `Runner.close_proc_stdin` is only going to be called once.)

This may not be a huge deal for Local but it's still crazy messy, and for any other Runner subclass (like in Fabric's Remote) it could potentially be far more inefficient if not outright harmful.

Need some way to track state such that we don't trigger a "dead thread" alarm while either exiting the loop after all, or at _least_ skipping over that method call after its first run.

**Comment by bitprophet** [comment]
Also, I went through some mental loops today over whether this should be a bugfix or a new feature (on account of adding the new Runner method) and ended up deciding on the latter. Plus the potential for extra buggy hijinx from this behavior change also strikes me as being at odds with the stability emphasis of bugfix branches.

**Comment by bitprophet** [comment]
Decided to simply solve the "don't close stdin a lot" problem with basic state-tracking logic, instead of fucking around with thread aliveness yet again. As noted, there's a sleep in that loop which should limit any problems with it hanging out til end-of-subprocess, even if we could technically close it down sooner.

**Comment by irdkwmnsb** [comment]
I found a similar problem in v2.2.0 on python3.11.4 on darwin.
When I run python3 -c "from invoke import run; run('cat')"
Everything I type is being written to stdout as expected. But, if I try to close the stdin sending EOF (ctrl + d) it is ignored and the process keeps running.
Here is where things get interesting.
If I move the process to background (ctrl + z), then bring it to foreground, the stdin is somehow reconnected back to the cat process. 
The behavior changes in two ways: the keystrokes are now being cat'ed only after line break, and sending EOF now closes the stdin and ends cat.

Any idea why this emerges?

Also, I found that when using pty=True piping to cat doesn't end the process. E.g. echo 123 | python3 -c "from invoke import run; run('cat', pty=True)" doesn't end. But I can end it if it's opened interactively with ctrl+z.
